### PR TITLE
Avoid repeated scanning of siblings array

### PR DESCRIPTION
**Pull request by chrmarti**
`glob.ts` part of #47187. I have left TODOs in `explorerViewer.ts` for the remaining work.

### CODE REVIEW

**Comment by bpasero** [comment]
@chrmarti this is a great change, thanks for doing it. I can already see a speedup from 500ms to 100ms for filtering the explorer with 2000 children in a folder. And that is, without using the idea of parsing the expression.

Now for that topic, can you help me out? I tried to cache the expression as a parsed thing but I get compile errors now. Here is the explorer filter with the changes to try out:

```typescript
// Explorer Filter
interface CachedParsedExpression {
	original: glob.IExpression;
	parsed: glob.ParsedExpression;
}

export class FileFilter implements IFilter {

	private hiddenExpressionPerRoot: Map<string, CachedParsedExpression>;
	private workspaceFolderChangeListener: IDisposable;

	constructor(
		@IWorkspaceContextService private contextService: IWorkspaceContextService,
		@IConfigurationService private configurationService: IConfigurationService
	) {
		this.hiddenExpressionPerRoot = new Map<string, CachedParsedExpression>();

		this.registerListeners();
	}

	public registerListeners(): void {
		this.workspaceFolderChangeListener = this.contextService.onDidChangeWorkspaceFolders(() => this.updateConfiguration());
	}

	public updateConfiguration(): boolean {
		let needsRefresh = false;
		this.contextService.getWorkspace().folders.forEach(folder => {
			const configuration = this.configurationService.getValue<IFilesConfiguration>({ resource: folder.uri });
			const excludesConfig = (configuration && configuration.files && configuration.files.exclude) || Object.create(null);

			if (!needsRefresh) {
				const cached = this.hiddenExpressionPerRoot.get(folder.uri.toString());
				needsRefresh = !cached || !objects.equals(cached, excludesConfig);
			}

			const excludesConfigCopy = objects.deepClone(excludesConfig); // do not keep the config, as it gets mutated under our hoods

			this.hiddenExpressionPerRoot.set(folder.uri.toString(), { original: excludesConfigCopy, parsed: glob.parse(excludesConfigCopy) } as CachedParsedExpression);
		});

		return needsRefresh;
	}

	public isVisible(tree: ITree, stat: ExplorerItem): boolean {
		return this.doIsVisible(stat);
	}

	private doIsVisible(stat: ExplorerItem): boolean {
		if (stat instanceof NewStatPlaceholder || stat.isRoot) {
			return true; // always visible
		}

		// Hide those that match Hidden Patterns
		const cached = this.hiddenExpressionPerRoot.get(stat.root.resource.toString());
		if (cached && cached.parsed(paths.normalize(path.relative(stat.root.resource.path, stat.resource.path), true), name => !!stat.parent.getChild(name))) {
			return false; // hidden through pattern
		}

		return true;
	}

	public dispose(): void {
		this.workspaceFolderChangeListener = dispose(this.workspaceFolderChangeListener);
	}
}
```

**Review by bpasero — APPROVED** [review]
I am approving since I pushed https://github.com/Microsoft/vscode/pull/53994/commits/852ca2a4161706fb1b74e3f3f9695c5aaa3e0711 to adopt this. Without the adoption I saw a slowdown, so I think we should only merge this when clients have adopted this.

**Comment by bpasero** [comment]
I pushed the additional changes to cache the expression and now I see around 20ms spend in filtering for a 2000 folder child 💯

**Comment by chrmarti** [comment]
@roblourens Let us know if you agree with the changes. Not much changes in Search. One thing I noted is that there is one call to fs.readdir() per match when there is a siblings clause due to the way things are wired up at https://github.com/Microsoft/vscode/blob/b74812161ec46c9011bba92d5ee1f4400aef0593/src/vs/workbench/services/search/node/ripgrepTextSearch.ts#L82 and https://github.com/Microsoft/vscode/blob/b74812161ec46c9011bba92d5ee1f4400aef0593/src/vs/workbench/api/node/extHostSearch.ts#L432. That hasn't changed with this change.

**Comment by roblourens** [comment]
LGTM. It should only call fs.readdir for matches that match the LHS of the sibling clause. Not every result, so I haven't bothered to cache those files, but I could do that.

**Review by roblourens — APPROVED** [review]

**Comment by isidorn** [comment]
@chrmarti great work, thanks for doing this

File: src/vs/base/common/glob.ts [hunk 1]
```diff
@@ -247,8 +247,8 @@ const T5 = /^([\w\.-]+(\/[\w\.-]+)*)\/?$/; 						   		// something/else
 
 export type ParsedPattern = (path: string, basename?: string) => boolean;
 
-// The ParsedExpression returns a Promise iff siblingsFn returns a Promise.
-export type ParsedExpression = (path: string, basename?: string, siblingsFn?: () => string[] | TPromise<string[]>) => string | TPromise<string> /* the matching pattern */;
+// The ParsedExpression returns a Promise iff hasSibling returns a Promise.
```
Comment (jens1o) [inline src/vs/base/common/glob.ts]: `if`

Comment (isidorn) [reply src/vs/base/common/glob.ts]: @jens1o it's a math thing, not an english thing

**Pull request merged**